### DISCUSSION TITLE
rk3566 mainline

### DISCUSSION THREAD

**Discussion by StuartIanNaylor** [Joshua-Riek/ubuntu-rockchip#482] [General]
Keeping to the premise of "This release is a work in progress and highly experimental. So please avoid bothering me with any issues encountered while using these images. Do not download if you are looking for a stable operating system with full hardware support." any chance you could try a rk3566 mainline Joshua.
I know Ambernic have mainlined a lot but no spoort matrix like we have with rk3588 but out of curiousity and if the rk3588 build could have a tweak to try and build a rk3566 image? 
Radxa do an Zero3 that is a serious alternative to the Pi02 and have no idea what the mainline state is like, but there is a lot shared and may also be near complete.

**Comment by Joshua-Riek** [comment]
I will contact Radxa for some dev boards of the Zero 3. Been very busy recently and have had little time for development,

**Comment by StuartIanNaylor** [comment]
On the forum do an @jack but could send you one if you wish.
I have no idea how it would go but could be interesting :)
They have a really interesting one with a Nic where have the poe is onboard and guess there is a hat in the wings that is little more than a cap filter board,
Dunno as many are curious.

**Comment by StuartIanNaylor** [comment]
There is also the Orange Pi3b boards and the Radxa Rock3, its strange how the images are a bit stinky.
I keep watching mainline as with rk3588 and the rk3566 has had many submissions also and curious to why no experimental exists.
Anyway I will have a look at 6.7 but think we still might be missing hdmi and panthor.

**Reply by Joshua-Riek** [reply]
I have the Zero 3 from Radxa, but I have not had much time to dive into it yet. 6.7 is missing both HDMI and panthor, however i will include a patch for HDMI in my tree.

**Reply by MarkhamLee** [reply]
I have the Orange Pi3B running the headless version of the "experimental zero support community" Armbian build:
* The Ubuntu build from Orange Pi is quite a bit slower than the Armbian one, ~2,200 on Sysbench primes @ 10k vs 3500 on Armbian, not sure if that was user error but overall I was really disappointed in its performance until I switched it to Armbian. 
* Geekbench scores were 205/533 vs 283/633 for a Raspberry Pi 4B 
* Everything worked out of the box w/ Armbian as far as LAN, Wi-Fi, HDMI output, etc. Even with a Pi I have to make a few tweaks to get both Wi-Fi and LAN working once I install Ubuntu. 
* I have it running as a node on my K3s setup, I have Zigbee2MQTT deployed on it  + four containers for monitoring my Kasa Smartplugs (via an "unofficial" python library) and a 5th container for monitoring air quality in my office. The only issue I've encountered are vs my the Raspberry Pis I have on the same cluster is that Rancher/K3s can't see the CPU utilization or RAM usage. 

Given the cost ($50.00) plus onboard NVME it will be quite the compelling value once a proper OS is put together, especially if I can get the GPIO pins working. 

TL/DR - I'd be happy to support putting together a mainline Ubuntu build for this board, whether it's testing or donating a device.

**Reply by StuartIanNaylor** [reply]
I really like the Radxa03w boards but just purchased the Opi3b as out of 4 images only Debian desktop runs on the R03w, which is a really bad start by Radxa.
Its sort of took me by surprise as Ambernic has posted much to mainline and there is so much crossover with the RK3588.
I have a tendency to try out Ubuntu as ArmNN only seems to run on Ubuntu , but also been wanting any image with working Vulkan for NCNN as with the Armv8.2 mat/mul vector instructions even the A55 gets quite a ML boost.

I don't want to create any extra work and was hoping the likes of Opi&Radxa might see what great ubuntu images these are and being rk3588 the difference and support needs might be minimal.
I also have a rk3568 as that pcie3.0 x2 is a great addition for a low cost NAS.

Hopefully its being taken as enthusiasm and not annoying but yeah just Ubuntu images for RK35xx devices would be wonderful and all that is needed.

**Comment by StuartIanNaylor** [comment]
PS someone has dona a miniarch rk3366 and also has the Opi3b https://github.com/warpme/miniarch/releases/download/20231112-6.6.8/MiniArch-20231112-6.7.0-board-rk3566.zero3w-SD-Image.img.xz
I will check that out.
One thing I have found is the rockchip kernel is turning off OPP But this code limits the frequency
kernel/drivers/soc/rockchip/rockchip_opp_select.c for https://cdn.discordapp.com/attachments/1163684917293625367/1192354977801314344/image.png?ex=65a8c61b&is=6596511b&hm=843140bef42bd1c51631d911f7c9cdef72274a6fb65e6cad73b400cfdd4b6890&

**Reply by Joshua-Riek** [reply]
I started to work on the rk3566 radxa zero 3, I need to figure out GPU drivers and WiFi, but after that, I should be able to make some Opi3b images as well; I have a board coming in the mail.

**Reply by StuartIanNaylor** [reply]
I am not all that bothered about the rk3566 G52‑2EE as its a single core and Mesa call it a G52-Lite as its a cut down G52.
Shame the Wifi driver messed up the networking on the ZeroW with wifi4 as it already looked much better than others as idle wattage was somewhere near normality

**Reply by Joshua-Riek** [reply]
I just pushed some code for the Orange Pi 3B, there is a build running now on GitHub actions. As for the Radxa Zero 3W, I need to re-work the wifi drivers and will take some time. I also have a Zero 3W model with the AIC8800 WiFi 6 chip coming in a few weeks.

**Reply by Joshua-Riek** [reply]
Device trees for the Radxa Zero 3 have been merged to the kernel, and I've confirmed that WiFi 4 works, but I still don't know about the WiFi 6 model. 

As for the Orange Pi 3B, I need to get bluetooth working.

**Reply by StuartIanNaylor** [reply]
Still waiting for the Opi3b, had the Armour case for some days just to annoy me :)
With the Zero3 Wifi4 have you got an image for that? I think I might have to get a 2gb/16gb Wifi6 from OkDo just to test as strangely what I think is one of the most competitive 1gb\8gbeMMC SBCs is not stocked anywhere apart from Allnet which currently has none.

**Reply by Joshua-Riek** [reply]
Here is the latest build. You will need to modify /boot/firmware/ubuntuEnv.txt and set the device tree to "rk3566-radxa-zero3-ap6212.dtb", then reboot the system. Also, you may need to enter "sudo ifconfig wlan0 up" when the system reboots for the wifi interface to start.

https://github.com/Joshua-Riek/ubuntu-rockchip/actions/runs/7611338452

**Reply by StuartIanNaylor** [reply]
Still trying to work out why the idle wattage is so high but at least you server image doesn't get as hot as the mainline arch image on the radxa forum.
I tried the desktop but maybe some groups dont go through that the oem installer would, not bothered as will take a look at the server version.
Thanks

**Reply by StuartIanNaylor** [reply]
I guess eth0 is built in?
But this time the network is not being killed and have a wlan0

**Reply by Joshua-Riek** [reply]
Eth0 is built in, this is because 2 device trees are needed for the hardware. The desktop image will likely have problems due to the limited ram on the system, services for tracker miner and a few others will keep crashing and failing.

**Reply by MarkhamLee** [reply]
Thank you, this is great news. I'm going to try the Orange Pi 3B build as soon as I get another NVME drive to install the OS on.

**Reply by StuartIanNaylor** [reply]
Desktop I thought it might be that the panfork repo is still being included as just removed it from /etc/apt, installed ubuntu-desktop-minimal on the server.
Its up and running now @ the Ubuntu Pro Sreen
I am presuming all in good time the Zero3W DTB
The wlan seems to be there but don't even know if set to onboard or external antenna as no networks found
That tracker & miner for nautilus I just disabled than unistall from the 1st google https://www.linuxuprising.com/2019/07/how-to-completely-disable-tracker.html
`systemctl --user mask tracker-extract-3.service tracker-miner-fs-3.service tracker-miner-rss-3.service tracker-writeback-3.service tracker-xdg-portal-3.service tracker-miner-fs-control-3.service`
`tracker3 reset -s -r`
Which hugely cuts background tasks but curiously idle wattage still high. Should really have a look at the opp table on the A55 on the rk3588 just out of curiosity.

**Reply by Joshua-Riek** [reply]
No need for panfork on rk3566, it should not be included on the desktop image, I will double check. 

Wlan works for me:
```
ubuntu@ubuntu:~$ sudo ifconfig wlan0 up
ubuntu@ubuntu:~$ sudo iwlist wlan0 scan
```

**Reply by Joshua-Riek** [reply]
Oh, I forgot to push the WiFi firmware; I will fix this shortly and start another build :)

**Reply by StuartIanNaylor** [reply]
With the desktop running its doesn't like that guess nmcli
```
ubuntu@ubuntu:~$ nmcli device show
GENERAL.DEVICE:                         wlan0
GENERAL.TYPE:                           wifi
GENERAL.HWADDR:                         50:41:1C:65:18:B6
GENERAL.MTU:                            1500
GENERAL.STATE:                          20 (unavailable)
GENERAL.CONNECTION:                     --
GENERAL.CON-PATH:                       --
IP4.GATEWAY:                            --
IP6.GATEWAY:                            --

GENERAL.DEVICE:                         enx8cae4cf4e378
GENERAL.TYPE:                           ethernet
GENERAL.HWADDR:                         8C:AE:4C:F4:E3:78
GENERAL.MTU:                            1500
GENERAL.STATE:                          10 (unmanaged)
GENERAL.CONNECTION:                     --
GENERAL.CON-PATH:                       --
WIRED-PROPERTIES.CARRIER:               on
IP4.ADDRESS[1]:                         192.168.1.9/24
IP4.GATEWAY:                            192.168.1.1
IP4.ROUTE[1]:                           dst = 0.0.0.0/0, nh = 192.168.1.1, mt = 100
IP4.ROUTE[2]:                           dst = 192.168.1.0/24, nh = 0.0.0.0, mt = 100
IP4.ROUTE[3]:                           dst = 192.168.1.1/32, nh = 0.0.0.0, mt = 100
IP6.ADDRESS[1]:                         fdd4:40f0:f40b:b400:8eae:4cff:fef4:e378/64
IP6.ADDRESS[2]:                         fe80::8eae:4cff:fef4:e378/64
IP6.GATEWAY:                            --
IP6.ROUTE[1]:                           dst = fdd4:40f0:f40b:b400::/64, nh = ::, mt = 100
IP6.ROUTE[2]:                           dst = fe80::/64, nh = ::, mt = 256

GENERAL.DEVICE:                         eth0
GENERAL.TYPE:                           ethernet
GENERAL.HWADDR:                         52:30:16:E0:DB:A9
GENERAL.MTU:                            1500
GENERAL.STATE:                          10 (unmanaged)
GENERAL.CONNECTION:                     --
GENERAL.CON-PATH:                       --
WIRED-PROPERTIES.CARRIER:               on
IP4.GATEWAY:                            --
IP6.GATEWAY:                            --
```
yep dunno why it says `GENERAL.STATE:                          20 (unavailable)` but prob something with starting with the server and doing the desktop install ontop so ignore for now.

**Reply by StuartIanNaylor** [reply]
Its no rush as the panfork repo I removed but left liujianfeng1994-ubuntu-rockchip-multimedia-jammy.list in /etc/apt/sources.list.d
I presume your build is doing the same and grabing panfork.
With firmware and is there an end of install where you can run OEM scripts?

https://www.linuxuprising.com/2019/07/how-to-completely-disable-tracker.html
systemctl --user mask tracker-extract-3.service tracker-miner-fs-3.service tracker-miner-rss-3.service tracker-writeback-3.service tracker-xdg-portal-3.service tracker-miner-fs-control-3.service
tracker3 reset -s -r
As disable as they just hog too much of the a55x4

**Reply by Joshua-Riek** [reply]
Could you put dmesg into a text file or upload to pastebin? Its likely missing firmware, i pushed it a few minutes ago but still building.

**Reply by StuartIanNaylor** [reply]
Coolio Iglesias, as if change "rk3566-radxa-zero3.dtb" which I presume is wifi 6 with firmware but change desktop to ""rk3566-radxa-zero3-ap6212.dtb" for wifi4 version and we have a working desktop with no panfrost that wifi connects to, for us with the ap6212.
I did notice something fail at the end of the OEM install script but didn't catch it.
I grabbed the next run 171

Still needed to 
`systemctl --user mask tracker-extract-3.service tracker-miner-fs-3.service tracker-miner-rss-3.service tracker-writeback-3.service tracker-xdg-portal-3.service tracker-miner-fs-control-3.service
tracker3 reset -s -r`
Maybe that what flashed on the cli as an error that I didn't catch.

**Reply by Joshua-Riek** [reply]
Thanks for the feedback, I'll look at disabling tracker miner by default. I still have no idea if wifi 6 works though

**Reply by StuartIanNaylor** [reply]
The Nautilus tracker & miner are just too much load on this SBC.
I don't have a wifi6, but will have the Opi3b soon hopefully, I will prob get a 2gb/16gb eMMC when in stock, but if you want guess I could get another 1gb wifi6 to test if you wish?

Tidy DTB disable does what you think disables internal
```
	ethernet@fe010000 {
		compatible = "rockchip,rk3568-gmac\0snps,dwmac-4.20a";
		reg = <0x00 0xfe010000 0x00 0x10000>;
		interrupts = <0x00 0x20 0x04 0x00 0x1d 0x04>;
		interrupt-names = "macirq\0eth_wake_irq";
		rockchip,grf = <0x35>;
		clocks = <0x23 0x186 0x23 0x189 0x23 0x189 0x23 0xc7 0x23 0xc3 0x23 0xc4 0x23 0x189 0x23 0xc8 0x23 0xac 0x23 0xab>;
		clock-names = "stmmaceth\0mac_clk_rx\0mac_clk_tx\0clk_mac_refout\0aclk_mac\0pclk_mac\0clk_mac_speed\0ptp_ref\0pclk_xpcs\0clk_xpcs_eee";
		resets = <0x23 0xec>;
		reset-names = "stmmaceth";
		snps,mixed-burst;
		snps,tso;
		snps,axi-config = <0x7e>;
		snps,mtl-rx-config = <0x7f>;
		snps,mtl-tx-config = <0x80>;
		status = "disabled";
		phy-mode = "rgmii";
		clock_in_out = "output";
		snps,reset-gpio = <0x81 0x10 0x01>;
		snps,reset-active-low;
		snps,reset-delays-us = <0x00 0x4e20 0x186a0>;
		assigned-clocks = <0x23 0x189 0x23 0x186>;
		assigned-clock-parents = <0x23 0x187>;
		assigned-clock-rates = <0x00 0x7735940>;
		pinctrl-names = "default";
		pinctrl-0 = <0x82 0x83 0x84 0x85 0x86>;
		tx_delay = <0x47>;
		rx_delay = <0x27>;
		phy-handle = <0x87>;
		phandle = <0x7d>;
```

**Reply by Joshua-Riek** [reply]
I have a wifi 6 board coming, but shipping is slow. What about the DTB? I have ethernet enabled, so there is no need for 3 device trees as it does not conflict with wifi.

**Reply by StuartIanNaylor** [reply]
It just seemed tidy as occasionally in Ubuntu it popped up unable to connect and thought it was that eth causing the desktop message.
As I guess on the 3W its enabled but no Wifi so another DTB.
From idle wattage to IO unavailabe but still enabled, to the green board led that is far too bright and flashes for anthing, I get the feeling the DTB's and Kernel config inherited is not at the same level of the RK3588(s) boards.
But yeah whatever is simplest for now, but disabling was also my curiousity to see if it would reduce the high idle wattage.

**Reply by StuartIanNaylor** [reply]
I got the Opi3b and the wifi driver is same but it was the 170 build (I think)
Apart from that all good.
Also somehow it seems to pull less watts than Zero3 ... ?
PS as I usually do with Opi boards I got the armor case/heatsink/shell or whatever you call it.
I did my usual mod of adding a thermal pad to the back of the cpu and backplate and with some m3 nuts raise the fan so it has something to blow through than choke.
Its probably the coolest running SBC I have with stress-ng failing to ever get to 45'C.
I may of copy/paste the wrong info on the tracker but the page with the info is good.
https://www.linuxuprising.com/2019/07/how-to-completely-disable-tracker.html
Interesting the schedutil govenor is working on the Opi3b and actually cloacks down to min freq and idle @ the wall approx 3.2watt.
Maybe that is what Zero3W should use as energy efficient for me is important with the 'small boards'
Usually I ignore schedutil but here it seems to be working fine, also though when I switched to ondemand the Opi3b still seems to take less watts.

**Reply by StuartIanNaylor** [reply]
With lower ram or even higher when trying to run LLMs having zram with a file swap as backup if zram hits uncompressable data.
The debian zram_config is awful 
https://github.com/StuartIanNaylor/zram-swap-config
I actually did a more advanced that can load / in an overlayfs with zram upper and merge down at shutdown.
I handed it to https://github.com/ecdye/zram-config as my MS just forgets...

**Reply by StuartIanNaylor** [reply]
Ps do you think https://ppa.launchpadcontent.net/liujianfeng1994/rockchip-multimedia/ubuntu jammy InRelease works for rk3566? I thought it was supposed to but looking at results I am wondering otherwise.

**Reply by Joshua-Riek** [reply]
Hmm I thought it worked, but maybe not.

**Reply by StuartIanNaylor** [reply]
No rush Joshua, I don't usually bother with desktop so even having to set up a Opi5 just to see how it runs there.
I am wondering if its running on cpu or the A55 just handling that causes it to max out on a core...
Even the Kodi interface cause considerable load on the rk3566

**Reply by StuartIanNaylor** [reply]
Nope I don't think it runs that great on the RK3588 but the CPU just handles it better.
/usr/bin/create-chromium-vda-vea-devices.sh
```
#!/bin/bash
case "$(cat /proc/device-tree/compatible)" in
	*rk3588*)
		{
			echo "type=dec"
			echo "codecs=VP8:VP9:H.264:H.265:AV1"
			echo "max-width=7680"
			echo "max-height=4320"
		} > /dev/video-dec0
		;;
	*rk3568*)
		{
			echo "type=dec"
			echo "codecs=VP8:VP9:H.264:H.265"
			echo "max-width=3840"
			echo "max-height=2160"
		} > /dev/video-dec0
		;;
	*)
		echo dec > /dev/video-dec0
esac
chown root:video /dev/video-dec0
chmod 0660 /dev/video-dec0
echo enc > /dev/video-enc0
chown root:video /dev/video-enc0
chmod 0660 /dev/video-enc0
```
Seems rk3566 is missing but as far as I know same as rk3568 but also udev rule seems purely for chromium

**Reply by Joshua-Riek** [reply]
I see, I can remove the ppa in future images.

**Reply by StuartIanNaylor** [reply]
Yeah I am thinking get rid, make the base image clean and let others install what they wish.
Even with Chromium on the rk3588 is it using hardware accel, lols want to see without now?
I guess just another case of
```
	*rk3566*)
		{
			echo "type=dec"
			echo "codecs=VP8:VP9:H.264:H.265"
			echo "max-width=3840"
			echo "max-height=2160"
		} > /dev/video-dec0
		;;
```
Is needed?
Dunno your call, for me I go vanilla and then write up a howto in the wiki sort of archlinux style.

**Reply by StuartIanNaylor** [reply]
PS Joshua as thinking about what I don't like about Armbian and have perference to yours and also why I really like Archlinux take of just super documentaion.
A entry for Zram but maybe take my horrid scripting and do with as you wish and add as a package in your ppa?
Same with the multimedia by Amazingfate maybe don't add it but have a Hardware Accel documentation entry showing how to add the PPA and install?

**Reply by StuartIanNaylor** [reply]
https://github.com/amazingfate/rockchip-multimedia-config/issues/2#issuecomment-1922639133

**Reply by StuartIanNaylor** [reply]
Another one is that hdmi audio doesn't play there is a difference in the alsa setting that with Ubuntu seems to show more than debian B6 which shows 2 objects only but strangely no volume but pulse audio in the desktop applies one ... ?
debianB6
```
	display-subsystem {
		compatible = "rockchip,display-subsystem";
		memory-region = <0x12 0x13>;
		memory-region-names = "drm-logo\0drm-cubic-lut";
		ports = <0x14>;
		devfreq = <0x15>;
		status = "okay";
		phandle = <0x11c>;

		route {

			route-dsi0 {
				status = "disabled";
				logo,uboot = "logo.bmp";
				logo,kernel = "logo_kernel.bmp";
				logo,mode = "center";
				charge_logo,mode = "center";
				connect = <0x16>;
				phandle = <0x11d>;
			};

			route-dsi1 {
				status = "disabled";
				logo,uboot = "logo.bmp";
				logo,kernel = "logo_kernel.bmp";
				logo,mode = "center";
				charge_logo,mode = "center";
				connect = <0x17>;
				phandle = <0x11e>;
			};

			route-edp {
				status = "disabled";
				logo,uboot = "logo.bmp";
				logo,kernel = "logo_kernel.bmp";
				logo,mode = "center";
				charge_logo,mode = "center";
				connect = <0x18>;
				phandle = <0x11f>;
			};

			route-hdmi {
				status = "okay";
				logo,uboot = "logo.bmp";
				logo,kernel = "logo_kernel.bmp";
				logo,mode = "center";
				charge_logo,mode = "center";
				connect = <0x19>;
				phandle = <0x120>;
			};

			route-lvds {
				status = "disabled";
				logo,uboot = "logo.bmp";
				logo,kernel = "logo_kernel.bmp";
				logo,mode = "center";
				charge_logo,mode = "center";
				connect = <0x1a>;
				phandle = <0x121>;
			};

			route-rgb {
				status = "disabled";
				logo,uboot = "logo.bmp";
				logo,kernel = "logo_kernel.bmp";
				logo,mode = "center";
				charge_logo,mode = "center";
				connect = <0x1b>;
				phandle = <0x122>;
			};
		};
	};
```

```
	hdmi@fe0a0000 {
		compatible = "rockchip,rk3568-dw-hdmi";
		reg = <0x00 0xfe0a0000 0x00 0x20000>;
		interrupts = <0x00 0x2d 0x04>;
		clocks = <0x21 0xe6 0x21 0xe7 0x21 0x193 0x32 0x02 0x21 0xde>;
		clock-names = "iahb\0isfr\0cec\0ref\0hclk";
		power-domains = <0x23 0x09>;
		reg-io-width = <0x04>;
		rockchip,grf = <0x33>;
		#sound-dai-cells = <0x00>;
		pinctrl-names = "default";
		pinctrl-0 = <0x95 0x96 0x97>;
		status = "okay";
		preset_max_hdisplay = <0x780>;
		preset_max_vdisplay = <0x438>;
		phandle = <0x11b>;

		ports {
			#address-cells = <0x01>;
			#size-cells = <0x00>;

			port@0 {
				reg = <0x00>;
				#address-cells = <0x01>;
				#size-cells = <0x00>;

				endpoint@0 {
					reg = <0x00>;
					remote-endpoint = <0x19>;
					status = "okay";
					phandle = <0x8a>;
				};

				endpoint@1 {
					reg = <0x01>;
					remote-endpoint = <0x98>;
					status = "disabled";
					phandle = <0x8e>;
				};
			};
		};
	};
```

```
	i2s@fe400000 {
		compatible = "rockchip,rk3568-i2s-tdm";
		reg = <0x00 0xfe400000 0x00 0x1000>;
		interrupts = <0x00 0x34 0x04>;
		clocks = <0x21 0x3f 0x21 0x43 0x21 0x39>;
		clock-names = "mclk_tx\0mclk_rx\0hclk";
		dmas = <0xaf 0x00>;
		dma-names = "tx";
		resets = <0x21 0x50 0x21 0x51>;
		reset-names = "tx-m\0rx-m";
		rockchip,cru = <0x21>;
		rockchip,grf = <0x33>;
		rockchip,playback-only;
		#sound-dai-cells = <0x00>;
		status = "okay";
		phandle = <0x11a>;
	};
```

```
	hdmi-sound {
		compatible = "simple-audio-card";
		simple-audio-card,format = "i2s";
		simple-audio-card,mclk-fs = <0x80>;
		simple-audio-card,name = "rockchip-hdmi0";
		status = "okay";
		phandle = <0x2e7>;

		simple-audio-card,cpu {
			sound-dai = <0x11a>;
		};

		simple-audio-card,codec {
			sound-dai = <0x11b>;
		};
	};
```
